vkubelet/vkubelet.go

```diff
@@ -32,123 +31,73 @@ type Server struct {
 	nodeName        string
 	namespace       string
 	k8sClient       *kubernetes.Clientset
-	taint           corev1.Taint
-	disableTaint    bool
-	provider        Provider
+	taint           *corev1.Taint
+	provider        providers.Provider
 	podWatcher      watch.Interface
 	resourceManager *manager.ResourceManager
 }
 
-func getEnv(key, defaultValue string) string {
-	value, found := os.LookupEnv(key)
-	if found {
-		return value
-	}
-	return defaultValue
+// Config is used to configure a new server.
+type Config struct {
+	APIConfig       APIConfig
+	Client          *kubernetes.Clientset
+	MetricsAddr     string
+	Namespace       string
+	NodeName        string
+	Provider        providers.Provider
+	ResourceManager *manager.ResourceManager
+	Taint           *corev1.Taint
 }
 
-// New creates a new virtual-kubelet server.
-func New(nodeName, operatingSystem, namespace, kubeConfig, provider, providerConfig, taintKey string, disableTaint bool, metricsAddr string) (*Server, error) {
-	var config *rest.Config
-
-	// Check if the kubeConfig file exists.
-	if _, err := os.Stat(kubeConfig); !os.IsNotExist(err) {
-		// Get the kubeconfig from the filepath.
-		config, err = clientcmd.BuildConfigFromFlags("", kubeConfig)
-		if err != nil {
-			return nil, err
-		}
-	} else {
-		// Set to in-cluster config.
-		config, err = rest.InClusterConfig()
-		if err != nil {
-			return nil, err
-		}
-	}
+type APIConfig struct {
+	Cert string
+	Key  string
+	Addr string
+}
 
-	if masterURI := os.Getenv("MASTER_URI"); masterURI != "" {
-		config.Host = masterURI
+// New creates a new virtual-kubelet server.
+func New(ctx context.Context, cfg Config) (s *Server, retErr error) {
+	s = &Server{
+		namespace:       cfg.Namespace,
+		nodeName:        cfg.NodeName,
+		taint:           cfg.Taint,
+		k8sClient:       cfg.Client,
+		resourceManager: cfg.ResourceManager,
+		provider:        cfg.Provider,
 	}
 
-	clientset, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		return nil, err
-	}
+	ctx = log.WithLogger(ctx, log.G(ctx))
 
-	rm, err := manager.NewResourceManager(clientset)
+	apiL, err := net.Listen("tcp", cfg.APIConfig.Addr)
 	if err != nil {
-		return nil, pkgerrors.Wrap(err, "error creating resource manager")
+		return nil, pkgerrors.Wrap(err, "error setting up API listener")
 	}
+	defer func() {
```

Comment:
this seems wrong. the apiL should not be closed in this method.

Reply:
This is only closing if the main function returns an error.